[JSC] Make Weak::~Weak inlined

Constellation opened the pull request:
#### 0c07f17913190779681015f86a52b642c03303dc
<pre>
[JSC] Make Weak::~Weak inlined
<a href="https://bugs.webkit.org/show_bug.cgi?id=256203">https://bugs.webkit.org/show_bug.cgi?id=256203</a>
rdar://108777077

Reviewed by Mark Lam.

weakClearSlowCase is not inlined while it is super tiny function because it is called from WebCore and
it is defined as JSC&apos;s non-inlined function, this is silly. We make Weak::clear inlined function so
that we can make Weak::~Weak fully inlined.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/Weak.cpp: Removed.
* Source/JavaScriptCore/heap/Weak.h:
(JSC::Weak::clear): Deleted.
* Source/JavaScriptCore/heap/WeakImpl.h:
(JSC::WeakImpl::WeakImpl):
(JSC::WeakImpl::clear):
(JSC::Weak&lt;T&gt;::clear):
* Source/JavaScriptCore/heap/WeakSet.h:
(JSC::WeakSet::deallocate):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/ScriptWrappable.h:
* Source/WebCore/bindings/js/ScriptWrappableInlines.h:

Canonical link: <a href="https://commits.webkit.org/263623@main">https://commits.webkit.org/263623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40d22128bb1b4a16a01155b61883a62aebaee0bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5308 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5263 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5381 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5831 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6819 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2919 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/11251 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4358 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4783 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4797 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6412 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4849 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5219 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4286 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5367 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4687 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1349 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8780 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5518 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/589 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5049 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1466 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/heap/WeakSet.h:
```diff
@@ -115,9 +115,9 @@ inline bool WeakSet::isTriviallyDestructible() const
+ALWAYS_INLINE void WeakSet::deallocate(WeakImpl* weakImpl)
 {
-    weakImpl->setState(WeakImpl::Deallocated);
+    weakImpl->clear();
```
This is a behavior change: the old code preserves the handle owner pointer, the new code nullifies it.  Why is this ok?

Constellation replied:
Discussed offline. This is intentional since WeakImpl gets deallocated. So owner is not touched anymore. This change avoids unnecessary load.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Make Weak::~Weak inlined
https://bugs.webkit.org/show_bug.cgi?id=256203
rdar://108777077

Reviewed by Mark Lam.

weakClearSlowCase is not inlined while it is super tiny function because it is called from WebCore and
it is defined as JSC's non-inlined function, this is silly. We make Weak::clear inlined function so
that we can make Weak::~Weak fully inlined.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/Weak.cpp: Removed.
* Source/JavaScriptCore/heap/Weak.h:
(JSC::Weak::clear): Deleted.
* Source/JavaScriptCore/heap/WeakImpl.h:
(JSC::WeakImpl::WeakImpl):
(JSC::WeakImpl::clear):
(JSC::Weak<T>::clear):
* Source/JavaScriptCore/heap/WeakSet.h:
(JSC::WeakSet::deallocate):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/ScriptWrappable.h:
* Source/WebCore/bindings/js/ScriptWrappableInlines.h:

Canonical link: https://commits.webkit.org/263623@main
